plugins/parsers/avro/parser.go

```diff
@@ -0,0 +1,238 @@
+		return nil, fmt.Errorf("cannot cast native interface {} to map[string]interface{}")
+	}
+
+	metricTime, err := p.parseTimestamp(nestedValue(deep[p.Timestamp]))
+	if err != nil {
+		return nil, err
+	}
+
+	fields := make(map[string]interface{})
+	tags := make(map[string]string)
+
+	for k, v := range p.DefaultTags {
+		tags[k] = v
+	}
+
+	for _, tag := range p.Tags {
+		if value, ok := deep[tag]; ok {
+			tags[tag] = fmt.Sprintf("%v", nestedValue(value))
+		} else {
+			p.Log.Infof("tag %s value was %v; not added to tags", tag, value)
+		}
+	}
+	fieldList := make([]string, len(p.Fields), (cap(p.Fields)+1)*2)
+	copy(fieldList, p.Fields)
+	if len(fieldList) == 0 { // Get fields from whatever we just unpacked
+		for k := range deep {
+			if _, ok := tags[k]; !ok {
+				fieldList = append(fieldList, k)
+			}
+		}
+	}
+	for _, field := range fieldList {
+		if value, ok := deep[field]; ok {
+			fields[field] = nestedValue(value)
+		} else {
+			p.Log.Infof("field %s value was %v; not added to fields", field, value)
+		}
+	}
+	var schemaObj map[string]interface{}
+
+	err = json.Unmarshal([]byte(schema), &schemaObj)
+	if err != nil {
+		return nil, err
+	}
+	if len(fields) == 0 {
+		return nil, fmt.Errorf("number of fields is 0; unable to create metric")
+	}
+	name := ""
+	if p.Measurement != "" {
+		name = p.Measurement
+	} else {
+		// get Measurement name from schema
+		nsStr, ok := schemaObj["namespace"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine namespace from schema %s", schema)
+		}
+		nStr, ok := schemaObj["name"].(string)
+		if !ok {
+			return nil, fmt.Errorf("could not determine name from schema %s", schema)
+		}
+		name = nsStr + "." + nStr
+	}
+	if name == "" {
+		name = p.MetricName
+	}
+	if name == "" {
+		return nil, fmt.Errorf("could not determine measurement name")
+	}
+	if p.DiscardArrays {
+		// Any non-scalars end up being a nil field
+		return metric.New(name, tags, fields, metricTime), nil
+	}
+	// But if we do it this way, we flatten any compound structures,
+	// including arrays.  Goavro is only going to hand us back
+	// arrays, not maps.
+	// The default (the separator string is empty) is equivalent to
+	// what streamreactor does.
+	sep := flatten.SeparatorStyle{
+		Before: "",
+		Middle: p.FieldSeparator,
+		After:  "",
+	}
+	flat, err := flatten.Flatten(fields, "", sep)
+	if err != nil {
+		return nil, err
+	}
+	return metric.New(name, tags, flat, metricTime), nil
+}
+
+func nestedValue(deep interface{}) interface{} {
+	if m, ok := deep.(map[string]interface{}); ok {
+		for _, value := range m {
+			return nestedValue(value)
+		}
+	}
+	return deep
+}
+
+func init() {
+	parsers.Add("avro",
+		func(defaultMetricName string) telegraf.Parser {
+			return &Parser{MetricName: defaultMetricName}
+		})
+}
+
+func (p *Parser) Init() error {
+	p.TimeFunc = time.Now
+	if p.Schema == "" && p.SchemaRegistry == "" {
+		err := fmt.Errorf("one of SchemaRegistry or Schema must be specified")
+		return err
+	}
```

Comment:
Please use the config options here
```suggestion
	if p.Schema == "" && p.SchemaRegistry == "" {
		return errors.New("one of 'schema_registry' or 'schema' must be specified")
	}
```

Furthermore, I think having both specified also should bring up an error shouldn't it?